Add ability to perform controlplane upgrades

invidian opened the pull request.

invidian commented:
Pushed some updates:
- Provided user feedback about the upgrade process:
  ```
  Ensuring that cluster controlplane is up to date.
  Ensuring controlplane component 'kube-apiserver' is up to date... Done.
  Ensuring controlplane component 'kubernetes' is up to date... Done.
  Ensuring controlplane component 'calico' is up to date... Done.
  Installing component 'metrics-server'...
  Succesfully installed component 'metrics-server'!
  ```
- Changes listening to randomize IP address rather than the port. This is because `kube-apiserver` allows to override what IP address will be advertised (be added to the endpoints of `kubernetes.default.svc` Service), but not the port. Because of firewalls, it's then easier to keep the port static and randomize the IP address we listen on.
  On platforms, where kube-apiserver needs to be exposed on all interfaces (e.g. Packet), we switch `kube-apiserver` in-cluster port to 7443 (on this port, `kube-apiserver` will listen on random local IP and `haproxy` will listen on Node IP), then in addition, `haproxy` will also listen on `0.0.0.0:6443` to expose API on all interfaces (including public ones).
  On platforms with only private network (e.g. AWS), ports setup remains the same.

invidian commented:
I've done some more testing based on @iaguis feedback.

It seems that `kube-apiserver` is consuming hell lot of memory, so squeezing 2 instances on a 2GB machine (default controller type on AWS) is problematic.

Also, it seems this patch breaks `metrics-server` for some reason, I need to investigate.

invidian commented:
>Also, it seems this patch breaks metrics-server for some reason, I need to investigate.

Fixed that, extra backslash broke aggregation flags.

invidian commented:
With following cluster configuration (left just important bits):
```
cluster "aws" {
  os_channel       = "edge"
  os_version       = "2387.99.0"
```

I was able to perform full cluster upgrade (including `kubelet`) and it took 2 minutes with images already pulled.

 However, on first attempt, I got some anomaly, where new `kubelet` instance registered as `localhost` instead of using hostname, even though in the container, `hostname` command was returning valid hostname. Perhaps we could use `--hostname-override` and specify hostname explicitly to avoid such issues.

invidian commented:
Alternative, simpler and more dummy approach: https://github.com/kinvolk/lokomotive/tree/invidian/controlplane-upgrades-alternative.

iaguis commented:
> Alternative, simpler and more dummy approach: https://github.com/kinvolk/lokomotive/tree/invidian/controlplane-upgrades-alternative.

I've been testing this approach and it seems to work. However, the API server flips between available and unavailable while the other kubernetes CP components are updating. Since we also retry for those, by the time they're installed the API server becomes stable, but this is a side effect of us waiting for the other CP components. Can we do a bit better here?

Otherwise I think this works and is simple so I'm happy with it. Also, as expected, HA updates work well too.

invidian commented:
>I've been testing this approach and it seems to work. However, the API server flips between available and unavailable while the other kubernetes CP components are updating. Since we also retry for those, by the time they're installed the API server becomes stable, but this is a side effect of us waiting for the other CP components. Can we do a bit better here?

Yes, this is what I will be working on today. I think I'll implement a check, which will make sure, that `kube-apiserver` `DaemonSet` is up to date and only then proceed. 

I also still need to check against bad flags.

invidian commented:
>I also still need to check against bad flags.

Unfortunately, as I said, pod checkpointer does not prevent or provide a way to recover if bad flag is specified for `kube-apiserver`, as there is no way to distinguish between the "regular" crash, which is part of the upgrade process and the real crash, if e.g. bad flag has been specified.

Additionally, bad flag is also persisted to `inactive-manifests`, so the recovery procedure is manual and difficult. It should look something like this:
- copy bad manifest from `inactive-manifests` directory somewhere to edit
- edit it to restore it to the good state
- move it to `manifests` directory
- wait for static API server to be created by kubelet
- delete existing `kube-apiserver` Pod and Daemonset to prevent it from being run, checkpointed etc.
- resolve bad flag issue
- run upgrade process again to ensure that right manifests are applied to the cluster
- move static `kube-apiserver` manifest from `manifests`

>Yes, this is what I will be working on today. I think I'll implement a check, which will make sure, that kube-apiserver DaemonSet is up to date and only then proceed.

It seems that in order to make that, we just need to wait until all replicas of DaemonSet are ready and then we can proceed. This also involves only `kube-apiserver`, so retry logic can be limited in scope to just this upgrade.

iaguis commented:
> It seems that in order to make that, we just need to wait until all replicas of DaemonSet are ready and then we can proceed. This also involves only `kube-apiserver`, so retry logic can be limited in scope to just this upgrade.

cool!

invidian commented:
I opened an upstream issue about SO_REUSEPORT to see if there is a chance to get the support for it in upstream: https://github.com/kubernetes/kubernetes/issues/88785

I also implemented this option for `kube-apiserver` and tested it to verify that it's working as expected. It is. See the issue for more details.

iaguis commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver-disruption.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: policy/v1beta1
```
I think this commit (https://github.com/kinvolk/lokomotive/pull/32/commits/4d112c0eab1f51cacf5ae454d32df4df87702f0f) needs a commit message a lot bigger. We should explain what steps we're taking to make updates reliable, link to the upstream SO_REUSEPORT issue and possible limitations.

invidian replied:
Sure, done: https://github.com/kinvolk/lokomotive/pull/32/commits/d1de22ac8481e48873c8da568843498d4fc2973d.

rata replied:
Can we mark this as resolved?

invidian replied:
I think so, as @iaguis approved the PR.

iaguis commented on pkg/components/util/install.go:
```diff
@@ -140,3 +135,28 @@ func InstallAsRelease(name string, c components.Component, kubeconfig string) er
+
+	// TODO: Add some logging implementation? We currently just pass an empty function for logging.
+	if err := actionConfig.Init(kube.GetConfig(kubeconfig, "", ns), ns, "secret", func(format string, v ...interface{}) {}); err != nil {
+		return nil, fmt.Errorf("failed initializing helm: %v", err)
```
Should we use `%w` here?
```suggestion
		return nil, fmt.Errorf("failed initializing helm: %w", err)
```

invidian replied:
Yes, done.

iaguis commented on pkg/components/util/install.go:
```diff
@@ -140,3 +135,28 @@ func InstallAsRelease(name string, c components.Component, kubeconfig string) er
+
+	_, err := histClient.Run(name)
+	if err != nil && err != driver.ErrReleaseNotFound {
+		return false, fmt.Errorf("failed checking for chart history: %v", err)
```
Should we use `%w` here?
```suggestion
		return false, fmt.Errorf("failed checking for chart history: %w", err)
```

invidian replied:
Yep.

iaguis commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -37,38 +61,42 @@ spec:
+          --requestheader-group-headers=X-Remote-Group \
+          --requestheader-username-headers=X-Remote-User \
+          {{ end -}}
+          --tls-private-key-file=/etc/kubernetes/secrets/apiserver.key
```
Should this be next to `--tls-cert-file=`?

invidian replied:
This is used to make handling last backslash easier. But agreed, it would be nice to have it paired. I'll move some other parameter to the bottom.

invidian replied:
Done. PTAL.

invidian commented on Makefile:
```diff
@@ -47,7 +47,7 @@ test: check-go-format run-unit-tests
 .PHONY: lint
 lint:
-	golangci-lint run --enable-all --disable=godox --max-same-issues=0 --max-issues-per-linter=0 --build-tags aws,packet,e2e,disruptive-e2e --new-from-rev=$$(git merge-base $$(cat .git/resource/base_sha 2>/dev/null || echo "master") HEAD) --modules-download-mode=$(MOD) --timeout=5m ./...
+	golangci-lint run --enable-all --disable=godox,lll --max-same-issues=0 --max-issues-per-linter=0 --build-tags aws,packet,e2e,disruptive-e2e --new-from-rev=$$(git merge-base $$(cat .git/resource/base_sha 2>/dev/null || echo "master") HEAD) --modules-download-mode=$(MOD) --timeout=5m ./...
```
I'll actually refactor the code a bit, so we don't have to disable it for now, as it's hard to justify why we should disable it.

invidian replied:
Done.

invidian commented on cli/cmd/cluster.go:
```diff
@@ -138,3 +145,41 @@ func clusterExists(ctxLogger *logrus.Entry, ex *terraform.Executor) bool {
+	}
+
+	if err := helmChart.Validate(); err != nil {
+		ctxLogger.Fatalf("chart is invalid: %v", err)
```
Fixed.

invidian commented:
I'm looking into commit by commit changes now and rebasing stuff to make sure all commits make sense.

invidian commented:
Rebased and resolved conflicts.

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
           readOnly: true
+        - name: data
+          mountPath: /run/kube-apiserver
+      - name: haproxy
```
I've read the full commit message (thank you very much for the big message explaining, I **really** appreciate it), I tried to follow the discussions linked there. Let me know if I'm missing something.

Also, thank you very much for trying so hard to fix this case. But now that the config to do it is here, I can't help to wonder... is it worth it? (Without this config, we wouldn't know how tricky it was or not, so thanks again for doing it! :))

The HAProxy as a sidecar only to manage avoid manual intervention when running a single controller and the update failed seems like a big cost to pay. We are expanding the attack surface (now HA Proxy bugs are very relevant to us), we need to care about keeping the version up to date and the helm whole chart config is more complicated.

The configuration is quite more complicated, actually (the 7443 port, the random IP to bind tricks, this not being a DaemonSet just to run multiple instances for the single node case).

Why do we care so much about upgrades to single controller nodes clusters that failed? Am I missing something?

invidian replied:
>Why do we care so much about upgrades to single controller nodes clusters that failed? Am I missing something?

Well, one argument for caring about it is, that this is what user with smaller cluster may want. Also, this way we don't need to have 2 different implementation for upgrades, but we just have one generic solution.

>The configuration is quite more complicated, actually (the 7443 port, the random IP to bind tricks, this not being a DaemonSet just to run multiple instances for the single node case).

Well, yes, but with help of the upstream, we can have exactly the same solution, but much simpler, and this is what I'm mainly hoping for.

We can perhaps re-evaluate it again.

invidian replied:
>What do you mean with exactly the same solution?

I mean we would have the exact same setup, with side container etc.

>If the HAProxy is removed (with a conditional in helm) for HA clusters, this same procedure will work, right? Or am I missing something?

If we combine it with `requiredDuringScheduling` pod anti affinity, then I _think_ yes. But if we do that, we could also revert the changes to the port setup. We would also needs to test scale up and scale down scenarios, when transitioning from single controller node to HA and back.

rata replied:
> > If the HAProxy is removed (with a conditional in helm) for HA clusters, this same procedure will work, right? Or am I missing something?
> 
> If we combine it with `requiredDuringScheduling` pod anti affinity, then I _think_ yes. But if 

Are you trying to imply that if we don't use that, then no?

Looking more at it, it is using `hostNetwork: true` (see [here](https://github.com/kinvolk/lokomotive/pull/32/files#diff-12be5db389a0fefe0038d10211a890b4R51)), so the behaviour should not change.

So, if we remove this sidecar that is totally unneeded for tha HA setup (I'd argue that HA setup being stable and secure is more important than single node controller) it should work and there is no need for them to be used in HA setups. Right?

HA setup is way more important than single node setup, IMHO. Therefore, IMHO, I wouldn't let a single node temporal hack leak into HA setup and add a bigger attack surface and pain points (like upgrading just to upgrade HAProxy, etc.)

> we do that, we could also revert the changes to the port setup. We would also needs to test scale up and scale down scenarios, when transitioning from single controller node to HA and back.

What? This is totally unrelated, we never supported going back and forth from single node to HA and is not supported now, right? If this is correct, this seems totally to decide on this.

If we want to add support for that, we can see the trade-off of having it then (and it might be as simple as removing the `if` to get the sidecar on all cases, if we *really* want that. I'd say may not even in that case want that, and just a simple migration maybe with a *temporal* HAProxy, and then remove it. Having a more complex architecture than needed, for your most important case, to support other not so important cases doesn't seem like a good trade-off in general, IMHO).

invidian replied:
I wonder how `requiredDuringScheduling` behaves with helm upgrades BTW. If you request 3 replicas and one controller node will fail to bootstrap, it may hang and roll back the upgrades then. We should probably test that.

Regarding the HAProxy for HA setup, I don't mind removing it. What do you think @iaguis ?

iaguis replied:
I agree with removing it if it works without it.

invidian replied:
_sticks "if/else" everywhere_

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+          echo "127.$(shuf -i 0-255 -n 1).$(shuf -i 0-255 -n 1).$(shuf -i 1-253 -n 1)" | tee /run/kube-apiserver/address
+        volumeMounts:
+        - name: data
+          mountPath: /run/kube-apiserver
```
If we decide to keep this sidecar container, I'd consider including it in the yaml only for the single node case. Having it doormant in the yaml, for the HA control plane case (the most common for production), doesn't seem like a good idea IMHO. I think an HA 3 controllers cluster is a very important setup, I'd not slip unneeded containers and bigger attack surface.

Or am I missing something and is not doormant?

rata replied:
Removed from the HA case, closing :)

rata commented on assets/lokomotive-kubernetes/bootkube/variables.tf:
```diff
@@ -139,3 +139,9 @@ variable "external_apiserver_port" {
 }
+
+variable "expose_on_all_ports" {
+  description = "Controls if kube-apiserver should listen on all interfaces on port 6443"
```
If false, it which interface will it listen? Looking at the variables file, I don't see a clear answer. So, this seems confusing, at least for me as a user :)

invidian replied:
I meant `NodeIP` of type `InternalIP` (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#nodeaddress-v1-core). Which you can see when running `kubectl get nodes -o wide`.

>what if the node has several private IPs? Which one or how it will pick it up?

I don't know how selection works then. We use `status.podIP` with `hostNetwork: true`. I guess that can be enforced by using `--node-ip` `kubelet flag.

invidian replied:
Added a comment to the code for it.

rata replied:
Awesome, thanks!

rata commented on cli/cmd/cluster.go:
```diff
@@ -188,6 +188,8 @@ func (c controlplaneUpdater) upgradeComponent(component string) {
 	update := action.NewUpgrade(actionConfig)
 
+	update.Atomic = true
+
```
Why not squash this in previous commit that added the code?

invidian replied:
Fixed.

rata commented on cli/cmd/cluster.go:
```diff
@@ -186,6 +187,25 @@ func (c controlplaneUpdater) upgradeComponent(component string) {
+			ctxLogger.Fatalf("Installing controlplane component failed: %v", err)
+		}
+	}
+
```
Again, why not squash this commit?

invidian replied:
Fixed.

rata replied:
Thanks!

rata commented on cli/cmd/cluster-install.go:
```diff
@@ -89,7 +89,7 @@ func runClusterInstall(cmd *cobra.Command, args []string) {
 		}
 
 		// TODO: Check what networking solution we use and update that.
-		for _, c := range []string{"kube-apiserver", "kubernetes", "kubelet", "calico"} {
```
idem squash

invidian replied:
Fixed.

rata replied:
Thanks!

rata commented on pkg/components/util/install.go:
```diff
@@ -35,8 +35,6 @@ func InstallComponent(name string, c components.Component, kubeconfig string) er
 
 // InstallAsRelease installs a component as a Helm release using a Helm client.
 func InstallAsRelease(name string, c components.Component, kubeconfig string) error {
-	actionConfig := &action.Configuration{}
```
(commenting here as no other easy way to comment on this commit) **IMHO** `Refactor *`is quite a bad commit title. TONs of things can be "refactors". Why not explain what is being done? Like "Add HelmActionConfig() to improve readability", "cleanup: move common code to Add HelmActionConfig()" (or whatever you think explains better what is being done)

invidian replied:
Fixed.

invidian commented:
Thanks for the review @rata! 

To address your comments:

>Added some comments, the HAProxy is what makes me doubt the most. Was this discussed earlier and I missed it? I'd like to know what others think too :). In any case, we can improve on it later (if we decide it's worth) if we need to merge this. So, really, no big blocker from my POV :-)

We had a short meeting about this last week on Thursday, as a side discussion to some other meeting, where I proposed this solution. We agreed together, that adding HAProxy to the mix is rather hacky and we will try doing the upgrades using pod checkpointer, which you can see the results in #72. Unfortunately, I found some blockers there, so as far as I'm aware, this solution is preferred for now.

#

Regarding the commits, I follow similar principles, so please clarify, so I can fix accordingly.

>Commits should be atomic (i.e. self-contained in the sense of not breaking tests, logically different changes in different commits)

I'm not 100% sure, but they should be. Do you have some particular example I should take care of?

>Titles should be meaningful ("fix", "refactor" would catch almost all otherwise) and, if possible, short to fit in the github UI

I always limit commit title to what `vim` suggest when doing `git commit`, I guess GitHub has some smaller limit. But well, GitHub is just one UI for `git` :smile: 

>Don't add code you later change, or add code you should have added before. Your development history, most of the times, is not important for the reviewers. So, splitting commits on when you find out of stuff is usually not helpful to anyone, just squash to previous commits (unless this is painful or something).

I agree with that and I was considering doing it before merging, but I left it like it is for now, so there is a bit more context of why things look the way they look.

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/variables.tf:
```diff
@@ -31,7 +31,7 @@ variable "worker_count" {
 variable "controller_type" {
   type        = string
-  default     = "t3.small"
+  default     = "t3.medium"
```
> Does it make sense to add a comment saying t3.small is not enough for single node clusters to handle upgrades but it is for HA clusters?

I guess so, will do.

>t3.small is still useful for HA controllers, right?

I wouldn't recommend it to be honest, as there not so much of free memory on the nodes still, but it should be OK.

invidian replied:
Done.

invidian commented on cli/cmd/cluster-install.go:
```diff
@@ -78,6 +78,18 @@ func runClusterInstall(cmd *cobra.Command, args []string) {
+	// TODO: Check what networking solution we use and update that.
+	for _, c := range []string{"kube-apiserver", "kubernetes", "kubelet", "calico"} {
+		cu.upgradeComponent(c)
+	}
```
Yeah, the commit message is not very clear. What I meant is, that if `cluster install` command is run on already installed cluster, we should make sure, that all controlplane components are up to date. Will clarify.

invidian replied:
Fixed.

rata replied:
Thanks!

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -24,6 +36,18 @@ spec:
     spec:
+      affinity:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
```
Yes, we could do that. I suggested it in https://github.com/kinvolk/lokomotive/issues/90, as we should then also apply it for other controlplane components, so it's slightly out of scope for this (already too big) PR.

rata replied:
Oh, thanks!

rata commented:
> Thanks for the review @rata!
> 
> To address your comments:
> 
> > Added some comments, the HAProxy is what makes me doubt the most. Was this discussed earlier and I missed it? I'd like to know what others think too :). In any case, we can improve on it later (if we decide it's worth) if we need to merge this. So, really, no big blocker from my POV :-)
> 
> We had a short meeting about this last week on Thursday, as a side discussion to some other meeting, where I proposed this solution. We agreed together, that adding HAProxy to the mix is rather hacky and we will try doing the upgrades using pod checkpointer, which you can see the results in #72. Unfortunately, I found some blockers there, so as far as I'm aware, this solution is preferred for now.

Ohh, thanks for the extra context! Now I understand more, and if it was agreed with others... I worry less about it :-)

> 
> Regarding the commits, I follow similar principles, so please clarify, so I can fix accordingly.
> 
> > Commits should be atomic (i.e. self-contained in the sense of not breaking tests, logically different changes in different commits)
> 
> I'm not 100% sure, but they should be. Do you have some particular example I should take care of?

Nope, just explaining all my personal taste so you had more context. Sorry I wasn't clear :)

> > Titles should be meaningful ("fix", "refactor" would catch almost all otherwise) and, if possible, short to fit in the github UI
> 
> I always limit commit title to what `vim` suggest when doing `git commit`, I guess GitHub has some smaller limit. But well, GitHub is just one UI for `git` smile

Yeah :)

> 
> > Don't add code you later change, or add code you should have added before. Your development history, most of the times, is not important for the reviewers. So, splitting commits on when you find out of stuff is usually not helpful to anyone, just squash to previous commits (unless this is painful or something).
> 
> I agree with that and I was considering doing it before merging, but I left it like it is for now, so there is a bit more context of why things look the way they look.

I think some things (like the big fat commit adding HAProxy) definitely helps :-)

invidian commented:
Updates:
- Created upstream PR with `SO_REUSEPORT` support: https://github.com/kubernetes/kubernetes/pull/88893.
- I squashed and re-organized commits, as suggested by @rata, I hope it's more condensed and consistent now. I also improved commit messages for existing commits.
- I added support for updating `flannel` as well. Currently the solution is not great, as we query for selected network solution via Terraform output, so quite large round trip, but this is the only solution we have right now. Regarding `kube-router`, I think it's not supported and should be removed. Correct me if I'm wrong.

invidian commented:
Tests are passing again.

iaguis commented on pkg/platform/packet/template.go:
```diff
@@ -251,4 +251,40 @@ provider "packet" {
 }
+
+# values.yaml content for all deployed charts.
+output "pod-checkpoiter_values" {
```
Typo:
```suggestion
output "pod-checkpointer_values" {
```

invidian replied:
Fixed. Sorry about that :pray:

rata commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection/templates/host-protection.yaml:
```diff
@@ -47,7 +47,7 @@ spec:
     protocol: TCP
     destination:
-      ports: [6443]
+      ports: [6443, 7443]
```
Hmm, why do we need this? I guess we only need 6443 open to the world, right?

invidian replied:
Ah yes, for the outside we only need 6443. I thought it includes internal NICs too.

rata replied:
It's for `bond0` the hostendpoint (at least now, it might change if we use the automatichostendpoint), so if it's on loopback it won't affect

rata replied:
Cool! One port not exposed! :)

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+        - /bin/sh
+        - -c
+        - |
+          echo "127.$(shuf -i 0-255 -n 1).$(shuf -i 0-255 -n 1).$(shuf -i 1-253 -n 1)" | tee /run/kube-apiserver/address
```
I'll try to make all this logic conditional, based on the number of controllers.

rata replied:
Cool! :-)

invidian commented on assets/lokomotive-kubernetes/bootkube/assets.tf:
```diff
@@ -33,6 +33,8 @@ resource "local_file" "kube-apiserver" {
     aggregation_client_cert = var.enable_aggregation == true ? base64encode(join(" ", tls_locally_signed_cert.aggregation-client.*.cert_pem)) : ""
     aggregation_client_key  = var.enable_aggregation == true ? base64encode(join(" ", tls_private_key.aggregation-client.*.private_key_pem)) : ""
+    replicas                = length(var.etcd_servers)
+    expose_on_all_ports     = var.expose_on_all_ports
```
I forgot to rename this parameter.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+          if [ -z $ADDRESS ]; then \
+            echo "ADDRESS not found" && \
+            exit 1; \
+          fi && \
```
I think so, yes. Will give it a try.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+          backend kube-apiserver
+            server 1 $ADDRESS:{{ template "port" . }} check" > /run/kube-apiserver/haproxy.cfg && \
+          echo "Connecting to $ADDRESS:{{ template "port" . }}" && \
+          until nc -zv $ADDRESS {{ template "port" . }}; do sleep 1; done && \
```
This is to make sure, that `kube-apiserver` container is up and running before we start HAProxy process, as we use TCP readiness probe and it will be "UP" as long HAProxy process starts, as HAProxy does not reject incoming connections, if no backend is available, but returns 503. Also yes, if for whatever reason that happens, then also pod checkpointer will kick in, and as HAProxy does not bind then yet, it will start just fine.

rata replied:
Okay, so this logic is to not show as ready the pod when the api server isn't ready?

If the api server crashes, this will still happen (as HAProxy will still be running), right?

rata replied:
I mean, that can only be "meaningful" in the initialization, as if the API server crashes the HAProxy will still be running and, if IIUC the comment: https://github.com/kinvolk/lokomotive/pull/32/files#diff-12be5db389a0fefe0038d10211a890b4R196-R199

HAProxy will not reject connections to API server even if it's down.

Is there some value on having this? IIUC, no readinessProbe is set in the non-HA case, and this infinite loop can only make sure it connected once, but what if it connects and API server crashes just immediately after?

I'm trying to see why this is useful in practice. Has it been useful to you when developing this?

invidian replied:
>Is there some value on having this? IIUC, no readinessProbe is set in the non-HA case, and this infinite loop can only make sure it connected once, but what if it connects and API server crashes just immediately after?

Then it will be restarted by the `kubelet`. But in general, `kube-apiserver` won't start accepting connections until it makes sure, that all flags etc are correct.

>I'm trying to see why this is useful in practice. Has it been useful to you when developing this?

We should have readiness probes, as they are used by helm to determine if the upgrade process succeeded. Without the readiness probe, upgrade is always assumed as successful, see #83 for more details.

rata replied:
Oh, that is a very good reason! Can you please add a comment above the readinessProbe? I can totally see my future self removing that (as it seems pointless as it only tests HAProxy). I don't want to remove that by mistake :-D

invidian replied:
Added:
```
          # We use TCP readiness probe and HAProxy does not reject connections if no backend is available,
          # so we wait until kube-apiserver is available here, so readiness of haproxy container represents
          # readiness of kube-apiserver, as kube-apiserver cannot have readiness probe set, as it listens
          # on random IP address.
```

rata replied:
It's missing the most important part:
```Readiness probe here tests that at least the apiserver was up at least once (because HAProxy waiting for apiserver to start), after that it will mean just that HAProxy is up. But this is needed for helm to perform atomic upgrades, helm can't do atomic upgrades in components without readinessProbe
```
?

invidian replied:
Added to readinessProbe comment.

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/variables.tf:
```diff
@@ -30,8 +30,14 @@ variable "worker_count" {
+  # having one single controlplane node, t3.small has not enough memory (2GB)
+  # to run more than one instance of kube-apiserver in parallel, so we need to use
+  # a bigger instance. With HA controlplane, t3.small should be fine, though for
+  # production setups, it's recommended to use bigger instances (like t3.medium), to
```
Removed (but just this one reference in this line). Or did you mean to remove all references to types?

rata replied:
I meant sust this file and just this comment :)

invidian commented on assets/lokomotive-kubernetes/bootkube/assets.tf:
```diff
@@ -33,6 +33,8 @@ resource "local_file" "kube-apiserver" {
     aggregation_client_cert = var.enable_aggregation == true ? base64encode(join(" ", tls_locally_signed_cert.aggregation-client.*.cert_pem)) : ""
     aggregation_client_key  = var.enable_aggregation == true ? base64encode(join(" ", tls_private_key.aggregation-client.*.private_key_pem)) : ""
+    replicas                = length(var.etcd_servers)
+    expose_on_all_ports     = var.expose_on_all_ports
```
>This variable, also, is not changing just that (where to listen). Is changing the ip and the port it will listen too (for the HAProxy tricks)

I'd argue about that. For external API access, port 6443 is always used. I'd say swap of in-cluster port from 6443 to 7443 is an implementation detail, at least for now.

>I'd go with a better variable name or maybe separate the concepts: one var to expose on all interfaces and another to choose the port to listen (i.e. two vars instead of one doing this black magic not obvious at all from the name)

Hm, I'm not sure why we would need 2 parameters, if we don't plan on changing them?

rata replied:
Well, an implementation detail that the cluster admin needs to be aware of. You can't manage a cluster that you don't know the components and how it is supposed to be working, IMHO.

And surely, we are changing them.

It's okay if you prefer to have one variable only for that too, IMHO. But it should be more clear what this does (and I guess some documentation in the chart variables would be awesome)

invidian replied:
It's now renamed to `expose_on_all_interfaces`. Is it sufficient? Can we mark it as resolved?

rata replied:
Sorry, I missed this. Let's close it!

invidian commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -271,6 +271,7 @@ storage:
           -A INPUT -p tcp --dport 2380 -j ACCEPT
           -A INPUT -p tcp --dport 2381 -j ACCEPT
           -A INPUT -p tcp --dport 6443 -j ACCEPT
+          -A INPUT -p tcp --dport 7443 -j ACCEPT
```
Yes, that should be conditional. Unfortunately we don't distinguish here between the interfaces types. Ideally we would only open 7443 to cluster internal network and leave 6443 open to the world.

rata replied:
Why to the internal network? Isn't all cluster traffic going to 6443 and just HAProxy sends it to localhost on port 7443?

invidian replied:
No, in-cluster traffic will be going either via port 6443 or port 7443, depending on what IP address is advertised into the endpoints of `kubernetes.default.svc`.

rata replied:
What? That is not deterministic and at least only one of those ports will be used?

Also, if in-cluster traffic can go to any IP... I'm kind of confused now with the whole picture of the HAProxy setup being used. When would in-cluster traffic change? Only when the pod is down and pod-checkpointer takes over?

rata replied:
Well, github is not showing here that the context changed (now this is conditional) :-/

In any case, can you please elaborate on when/why in-cluster traffic can go to either port?

invidian replied:
So, this is rather not very well known feature of kube-apiserver as far as I know, but I'll try to keep it short. `--advertise-address` flag of `kube-apiserver` defines which IP address will be added as an endpoint to `kubernetes.default.svc` service, where in-cluster traffic to API will be going. It's uniqueness though it's based on IP address and not on IP+port combination. This is why we randomize the IP address to listen on and we set `--advertise-address` to Node IP, to keep it deterministic. If we would randomize the port, we would have to open all the ports in the range on the firewall to get it to work.

So, as uniqueness of `Endpoint` in `kubernetes.default.svc` `Service` is based on IP address, the address with port 6443 will be added first, as this is how bootstrap `kube-apiserver` binds. This port is later replaced by HAProxy. However, if you delete `kubernetes.default.svc` service (`kubectl delete service -n default kubernetes`), then `kube-apiserver` will re-create with existing listening port, which will be `7443` and then the traffic flows directly into the `kube-apiserver` process. The removal should not happen under regular operation, but we should be prepared for it.

This all would be much simpler, if `kube-apiserver` would be capable of advertising endpoints based on IP + port.

rata replied:
Oh, I see... but not sure. The service can specify a port, if we do specify it in the service... wouldn't we just remove the need to use 7443 port for in-cluster communication and can rely only on 6443?

In any case, I don't think these tricks should be understood by one person only. If that happens, the code is not written properly with all the comments it needs for other developers to work on it. IMHO, it might be complicated, but we need to be explicit about hacks, limitations and workarounds and it shouldn't be hidden in a solved github comment in some "past" PR (i.e. if we merge with only this conversation as the only guide)

rata replied:
I sometimes think I'm following it and sometimes doubt. We can create an issue or add a comment in a following PR explining this, maybe?

invidian replied:
>Oh, I see... but not sure. The service can specify a port, if we do specify it in the service... wouldn't we just use the port in the service and remove the need to use 7443 port for in-cluster communication and can rely only on 6443?

This is not about the service port, but the endpoints port. The service port is always 443, but it gets translated into endpoint port by kube-proxy.

>In any case, I don't think these tricks should be understood by one person only. If that happens, the code is not written properly with all the comments it needs for other developers to work on it. IMHO, it might be complicated, but we need to be explicit about hacks, limitations and workarounds and it shouldn't be hidden in a solved github comment in some "past" PR (i.e. if we merge with only this conversation as the only guide)

I'll add some documentation.

invidian replied:
I added some comments into the file to describe this change.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+
+          {{- end }}
+          backend kube-apiserver
+            server 1 $ADDRESS:{{ template "port" . }} check" > /run/kube-apiserver/haproxy.cfg && \
```
I'm not sure how moving it would make it "easier", but I think it make sense from the point of separating "setup" step from "runtime" step.

invidian replied:
Moved to init container.

rata replied:
Yeah, sorry, I meant that :-)

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -37,38 +61,42 @@ spec:
+        - -c
+        - |
+          set -x && \
+          exec /hyperkube \
```
`sh` is needed for `--bind-address=$(cat /run/kube-apiserver/address) \`.

rata replied:
And now this is used for the non HA control plane case only. Cool! Thanks!

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+          {{- end }}
+          backend kube-apiserver
+            server 1 $ADDRESS:{{ template "port" . }} check" > /run/kube-apiserver/haproxy.cfg && \
+          echo "Connecting to $ADDRESS:{{ template "port" . }}" && \
```
Explained in https://github.com/kinvolk/lokomotive/pull/32#discussion_r389535650.

rata replied:
Thanks! Let's wait for it to be up at least once before helm consider this succesful

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+            echo "ADDRESS not found" && \
+            exit 1; \
+          fi && \
+          echo "defaults \
```
I simplified it now.

rata replied:
It was buggy before? Just to make sure I understand :)

invidian replied:
Not, everything was running fine. It was just a leftover from fighting to get it to work right. using `- |` in YAML makes it not needed.

rata replied:
Ohh, okay! :)

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +109,67 @@ spec:
+            echo "ADDRESS not found" && \
+            exit 1; \
+          fi && \
+          echo "defaults \
```
I simplified it now.

rata replied:
I see it exactly the same regarding this comment. Am I missing something?

invidian replied:
Ah, I see what you mean now. Well, I think `"` is fine to be multiline, it plays well with keeping it in YAML. I'd prefer not to change it.

rata replied:
Yeah, but as I said, adding `"` in the configuration will break it. Maybe a comment to make sure about that is worth it (if you don't want to use the `cat` alternative commonly used)?

invidian replied:
I'm not sure if like the new syntax, but I will push it. It looks like this now:
```sh
          set -xe
          export ADDRESS="127.$(shuf -i 0-255 -n 1).$(shuf -i 0-255 -n 1).$(shuf -i 1-253 -n 1)"
          echo $ADDRESS > /run/kube-apiserver/address
          cat <<EOF > /run/kube-apiserver/haproxy.cfg
          defaults
            # Do TLS passthrough
            mode tcp
            # Required values for both frontend and backend
            timeout connect 5s
            timeout client 30s
            timeout client-fin 30s
            timeout server 30s
            timeout tunnel 21d

          frontend kube-apiserver-internal
            bind $POD_IP:{{ template "port" . }}
            default_backend kube-apiserver

          {{- if .Values.apiserver.exposeOnAllInterfaces }}
          frontend kube-apiserver-external
            bind 0.0.0.0:6443
            default_backend kube-apiserver

          {{ end }}
          backend kube-apiserver
            server 1 $ADDRESS:{{ template "port" . }} check"
          EOF
          cat /run/kube-apiserver/haproxy.cfg
```

rata replied:
I'm ok with both, just wanted to avoid "surprises" if you use a `"`.

invidian replied:
Okay, so it's updated now. Can we resolve?

rata replied:
Sorry!

invidian commented:
@rata I just pushed my work from today, addressing all your comments. Please take a look again. I cannot give 100% that it works on all combinations, as I didn't have enough time to test it yet, but it should mostly work.

rata commented:
@invidian awesome!! Will have a look! Thanks again! :)

invidian commented:
Tested scenarios:
- HA on Packet + upgrades
- standalone controller on Packet + upgrades
- standalone controller on AWS + upgrades
- HA on AWS + upgrades

@iaguis please take a look again then

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +173,78 @@ spec:
+        - -c
+        - |
+          set -xe
+          export ADDRESS="127.$(shuf -i 0-255 -n 1).$(shuf -i 0-255 -n 1).$(shuf -i 1-253 -n 1)"
```
`shuf` seems to be part of coreutils. Here we are using the alpine image, that (in a quick search) doesn't install that package, but seems to be a dependency of another installed package: https://github.com/docker-library/repo-info/blob/master/repos/haproxy/local/2.1.3-alpine.md

The debian image (way bigger, probably) does [install coreutils](https://github.com/docker-library/repo-info/blob/master/repos/haproxy/local/2.1.3.md#dpkg-source-package-coreutils830-3).

So, using this command here seems a little fragile (might break when upgrading) making the whole control plane startup to fail if a later image doesn't include coreutils (package dependencies change or the package that has this as dependency is not installed anymore).

Am I missing something? Do we want to depend on this although it seems fragile?

invidian replied:
I think saying it's fragile is exaggeration here. It works right now and I don't think suddenly HAProxy would stop depending on `coreutils`. Also see the progress on upstream issue regarding SO_REUSEPORT. I really hope we won't keep this solution for long.

rata replied:
Ohh, very nice, it seems to be making it's way forward upstream! Awesome, and congrats too! :)

Maybe create a tech-debt issue to either remove this (when SO_REUSEPORT is ready, hopefully in k8s 1.19) or make it more reliable? If you agree to create such an issue, an easy alternative can be this (works on `busybox` sh that alpine uses): `echo $((RANDOM%255))` (or +1 too to mimic the last octet here)

invidian replied:
OK, changed.

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +173,78 @@ spec:
+
+          frontend kube-apiserver-internal
+            bind $POD_IP:{{ template "port" . }}
+            default_backend kube-apiserver
```
Not sure I follow... this is to handle the case of a single controller node configured without HAProxy?

I thought we didn't allow those kind of installations at the lokoctl level. Do we want to support those setups or just have this "single node, then eat HAProxy" to simplify the yamls?

invidian replied:
I don't understand.

>Not sure I follow... this is to handle the case of a single controller node configured without HAProxy?

We don't have a case of single controller node without HAProxy. With single controller node, we deploy HAProxy to handle SO_REUSEPORT for us.

rata replied:
Okay... so, I think I'm confused on why we need two different listeners on HAProxy then?

Sorry, I think I'm not getting the full picture or kind of tired :D

invidian replied:
>Sorry, I think I'm not getting the full picture or kind of tired :D

I just realized that there is so many moving pieces to understand for that, so I understand the confusion :smile:. I Added a comment at the top of the file explaining, why we need to listen on 2 IP addresses.

rata replied:
I'm kind of lost to which comment do you refer to. But thanks! Maybe the comment is this: https://github.com/kinvolk/lokomotive/pull/32/files#diff-12be5db389a0fefe0038d10211a890b4R1-R8?

I think I got it now, more tricky than I thought originally! :-D

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -37,6 +90,44 @@ spec:
       - name: kube-apiserver
         image: {{ .Values.apiserver.image }}
         command:
+        {{- if eq (int .Values.apiserver.replicas) 1 }}
```
I'd add a big fat comment explaining that we use different things for 1 replica vs more replicas, and explain HAProxy hacks maybe here too?

I think it should be explained somewhere in the code. The code is quite long because of this, the user/devs reading it can use a hint here to understand these big and very different configurations

invidian replied:
I'm not sure if putting a big comment in already big YAML make a lot of sense, `git` history has the reasoning. Considering, that it takes a lot of lines to describe why we do it, I wouldn't put the copy in the YAML, in the commit message, in the PR and on Slack :) I can add a short reference if you find it valuable.

invidian replied:
We could perhaps rename this parameter to something like `controllers_count` or something to make it more self-explainatory.

rata replied:
I think a reference (here or in other part of the code) is very valuable. It's very important for anyone reading that file, IMHO. It's a non trivial architecture and we need to describe it somehow, IMHO, so people can do reasonable modifications (and not things that will fail because they don't understand the architecture)

invidian replied:
Added some short explanation.

rata replied:
Maybe not pushed? Or where is it? :)

invidian replied:
```diff
--- a/assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml
+++ b/assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml
@@ -1,5 +1,28 @@
+# On platforms like AWS, it's enough for kube-apiserver to listen on Node InternalIP. On other
+# like Packet, we need to listen on all interfaces to be able to expose kube-apiserver to the
+# internet. As we cannot listen on specific IP address and on 0.0.0.0 on the same port when using
+# SO_REUSEPORT, kube-apiserver process must listen on different port than HAProxy does. So if we
+# need to expose kube-apiserver on all interfaces, kube-apiserver will be listening on random
+# 127.0.0.0/8 IP address on port 7443, but will advertise Node InternalIP with port 7443 and HAProxy
+# will be listening on both Node InternalIP on port 7443 and on 0.0.0.0:6443 for both internal and
+# external traffic.
+# See commit messages and comments :https://github.com/kinvolk/lokomotive/pull/32 for more details.
+{{ define "port" }}
+{{- if and .Values.apiserver.exposeOnAllInterfaces (eq (int .Values.apiserver.replicas) 1) -}}
+7443
+{{- else -}}
+6443
+{{- end -}}
+{{ end }}
 apiVersion: apps/v1
+# If there is just one controller node, we want to use Deployment to be able to run 2 kube-apiserver
+# pods on a single node at a time, to provide graceful upgrades.
+# See commit messages and comments in https://github.com/kinvolk/lokomotive/pull/32 for more details.
+{{- if eq (int .Values.apiserver.replicas) 1 }}
+kind: Deployment
+{{- else }}
 kind: DaemonSet

```

rata replied:
Ohh, sorry! Thanks!

rata commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -271,6 +271,9 @@ storage:
           -A INPUT -p tcp --dport 2381 -j ACCEPT
           -A INPUT -p tcp --dport 6443 -j ACCEPT
+          %{~ if controllers == "1" ~}
+          -A INPUT -p tcp --dport 7443 -j ACCEPT
```
This seems tricky as calico is not overriding (yet?) these rules, but this rule is not on calico global network policies. If we need the rule during the cluster lifecycle (as I understand from https://github.com/kinvolk/lokomotive/pull/32/files#r390412048), I think we should re-add it there. But, if it's only for internal traffic, add it only for internal traffic.

Not sure, though, if it makes sense to duplicate the rule in calico and here... Maybe we want in case there are issues with calico?

invidian replied:
Keep in mind, that kube-apiserver will only listen on host internal IP, not on the external one, so this rule is no-op for external IP address. I'd say it's fine'ish for now

rata replied:
Should we create a tech-debt issue to follow up on this later?

invidian replied:
Yes. I'll create it.

rata replied:
Thanks, please link and mark as resolved :)

invidian replied:
Created #137.

rata commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:
```diff
@@ -104,6 +104,7 @@ data "template_file" "controller-configs" {
     ssh_keys              = jsonencode(var.ssh_keys)
     k8s_dns_service_ip    = cidrhost(var.service_cidr, 10)
     cluster_domain_suffix = var.cluster_domain_suffix
+    controllers           = var.controller_count
```
`controllers_nr` or  `controllers_count` seems more clear, IMHO

invidian replied:
Maybe `controller_count` to keep it consistent with Terraform?

invidian replied:
Renamed.

rata commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:
```diff
@@ -104,6 +104,7 @@ data "template_file" "controller-configs" {
     ssh_keys              = jsonencode(var.ssh_keys)
     k8s_dns_service_ip    = cidrhost(var.service_cidr, 10)
     cluster_domain_suffix = var.cluster_domain_suffix
+    controllers           = var.controller_count
```
The node_private_cidr can be added here to use in the iptables rule, right? I mean, if we need only node traffic to the 7443 port

invidian replied:
Yes, we could do that, but as said, we only listen on private IP address, so that would be no-op anyway for external traffic. I'm not sure if I want to make it more complex at this point.

rata replied:
Oh, okay!

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,42 @@
+{{ define "port" }}
+{{- if and .Values.apiserver.exposeOnAllInterfaces (eq (int .Values.apiserver.replicas) 1) -}}
```
It's the `helm` syntax for expressing `if foo && bar`.

rata replied:
Ohhh, never even realized! Tried to search if a reasonable syntax is there quickly but didn't find any :-(

Thanks!

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,42 @@
+{{- else -}}
+6443
+{{- end -}}
+{{ end }}
```
One closes the if, another one closes the `define` block.

rata replied:
Oh, didn't know define needs ends too. Sorry :see_no_evil:

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver-disruption.yaml:
```diff
@@ -0,0 +1,15 @@
+metadata:
+  name: kube-apiserver
+spec:
+  {{- if eq (int .Values.apiserver.replicas) 1 }}
```
Ah yes, that's the leftover. I later found out, that PodDisruptionBudget has no effect on DaemonSet and it actually produces a warning event on the cluster.

rata replied:
Awesome! :)

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,42 @@
 spec:
+  {{- if eq (int .Values.apiserver.replicas) 1 }}
+  replicas: 1
+  {{- end }}
```
If replicas is not 1, then DaemonSet is used, so `replicas` field shouldn't be there.

rata replied:
It's a daemon set on the other case now :man_facepalming:

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,42 @@
     type: RollingUpdate
     rollingUpdate:
+      {{- if eq (int .Values.apiserver.replicas) 1 }}
+      maxUnavailable: 0
```
This is how rolling update works, it adds additional replica and makes sure it becomes ready before removing old one.

rata replied:
:man_facepalming:

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,42 @@
       k8s-app: kube-apiserver
+  {{- if eq (int .Values.apiserver.replicas) 1 }}
+  strategy:
+  {{- else }}
```
Please see the rest of the template, `DaemonSet` has `updateStrategy` field, while `Deployment` has just `strategy`, but they point to the same block type, so further the content is the same.

rata replied:
Just realized about it one minute ago! Sorry for the noise, and thanks!

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -81,6 +173,78 @@ spec:
+          echo "Connected"
+          # From https://github.com/docker-library/haproxy/blob/master/Dockerfile-debian.template#L70
+          exec haproxy -f /run/kube-apiserver/haproxy.cfg
+        readinessProbe:
```
I see only the non HA case has a readinessProbe? According to https://github.com/kinvolk/lokomotive/pull/32#discussion_r390459035 don't we want to add one in the HA case too? If that is out of scope and tracked in some other place, fine by me. But if we can add a commit with this very same readinessProbe in the PR, seems on topic for me too

invidian replied:
I think we can add it for HA case too to make upgrade process smoother.

invidian replied:
I added it, but didn't test yet.

rata replied:
Thanks!

invidian replied:
Tested now. Seems to work fine, though it doesn't play very well with pod checkpointer, as it kicks in right after the pod is descheduled when upgrading the DaemonSet.

rata replied:
Ohh, the readinessProbe doesn't bode well? So maybe just removing it is better?

invidian replied:
It's better with `readinessProbe`, as it then upgrade process waits until new `kube-apiserver` is running correctly.

iaguis replied:
I wouldn't block the PR on this, create an issue and we can fix later.

rata replied:
Agreed. At this point, it seems better to merge and improve later for most of the remaining things (as long as basic tests we care about work fine, of course :))

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver-disruption.yaml:
```diff
@@ -0,0 +1,15 @@
+{{- if eq (int .Values.apiserver.replicas) 1 }}
```
Addressed.

rata replied:
Thanks!

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,53 @@
 apiVersion: apps/v1
+# If there is jsut one controller node, we want to use Deployment to be able to run 2 kube-apiserver
+# pods on a single node at a time, to provide graceful upgrades.
+# See https://github.com/kinvolk/lokomotive/pull/32 for more details.
```
This link doesn't seem very useful, maybe point to a specific comment or something?

invidian replied:
I mentioned to look for commit messages and comments. Does this make more sense?

rata replied:
I think mentioning the commit that explains this issue in a lot detail is way better (as I commented here: https://github.com/kinvolk/lokomotive/pull/32#discussion_r390502037). Does it make sense for you too?

invidian replied:
Ok, will do.

invidian replied:
Done.

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,54 @@
+{{- end -}}
+{{ end }}
 apiVersion: apps/v1
+# If there is jsut one controller node, we want to use Deployment to be able to run 2 kube-apiserver
```
s/jsut/just/

invidian replied:
Oops, fixed, sorry

rata replied:
Thanks!

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml:
```diff
@@ -1,20 +1,54 @@
 apiVersion: apps/v1
+# If there is jsut one controller node, we want to use Deployment to be able to run 2 kube-apiserver
+# pods on a single node at a time, to provide graceful upgrades.
+# See commit messages and comments in https://github.com/kinvolk/lokomotive/pull/32 for more details.
```
Again, I'm not a fan of the "go and spend 10 hours reading" instead of providing a link to the specific relevant comment

invidian replied:
I don't know where to point there, as there is to many aspects to take into account. Maybe I should have a design doc describing that? :joy: CC @iaguis

rata replied:
What about pointing to this commit: https://github.com/kinvolk/lokomotive/pull/32/commits/04f9737a4947d89e51f9ae066a319c92aaa5c27a (of course, hash might change on ammends, just update it as a very last thing?)

I think there explains the problem in detail quite well

invidian replied:
Ok, will do.

invidian replied:
Done. I hope I don't forget to update it before merging, if there are further changes :see_no_evil:

ipochi commented on pkg/components/util/helm_test.go:
```diff
@@ -32,3 +33,133 @@ func TestRenderChartBadValues(t *testing.T) {
+	}
+}
+
+func TestCharatFromManifestsRemoveNamespace(t *testing.T) {
```
Charat ? You mean Chart ?

invidian replied:
Oh, nice catch, fixed!

invidian commented:
I'm testing upgrades of kube-apiserver run as DaemonSet on HA Controlplane. It seems, that when kube-apiserver pod gets descheduled, pod checkpointer kicks in on every node, which triggers crashloop for a minute for every node. During this time, I also get connection refused often when running kubectl get pods.

With Deployment approach, there is 0 errors :see_no_evil:

rata commented:
@invidian that is weird, if I run a `kubectl set image` on an "old" (pre helm controlplane charts), it works just fine during the upgrad (kubectl does retries). And kubectl even doesn't fail even if one out of 3 controllers is down, it just takes more time (in the tests I did in the past).

How are you trying?

invidian commented:
>How are you trying?

I'm running `lokoctl cluster install` with changes hyperkube image version, which will trigger contorlplane upgrade. In the meanwhile, I just run `kubectl get pods` from time to time and I see, that it fails sometimes.

invidian commented:
For future self, with SO_REUSEPORT support in `kube-apiserver`, this patch needs to be applied: https://gist.github.com/invidian/d77dca57bcf3dbe49fa9b18330b5366a.
